[popover] Integrate with anchor positioning

anttijk opened the pull request:
#### 6bea8210e2b59ee2e34de9cc7d860677cb6086ee
<pre>
[popover] Integrate with anchor positioning
<a href="https://bugs.webkit.org/show_bug.cgi?id=284226">https://bugs.webkit.org/show_bug.cgi?id=284226</a>
<a href="https://rdar.apple.com/141094328">rdar://141094328</a>

Reviewed by Alan Baradlay and Tim Nguyen.

See <a href="https://github.com/whatwg/html/pull/10728.">https://github.com/whatwg/html/pull/10728.</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/popover-implicit-anchor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/popover-implicit-anchor.html:

Test against the actual button position so button margins from appearance don&apos;t affect the results.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setInvokedPopover):

Invalidate the element style so the we update the isPopoverInvoker style bit.

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::isPopoverInvoker const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setIsPopoverInvoker):

Add a bit. We use it to determine of a box is an implicit anchor.
Making this a style bit allows anchor registration and updates work correctly
without any new mechanisms.

* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::findImplicitAnchor):

The implicit anchor for a popover is its invoker.
Factor into a function.

(WebCore::Style::findLastAcceptableAnchorWithName):
(WebCore::Style::AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement):
(WebCore::Style::AnchorPositionEvaluator::isAnchor):
(WebCore::Style::AnchorPositionEvaluator::isImplicitAnchor):

A popover invoker is an implicit anchor.
Factor into a function.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Set isPopoverInvoker bit for invokers.

Canonical link: <a href="https://commits.webkit.org/295995@main">https://commits.webkit.org/295995@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12d4116b6432f1155332dab15de11a3a6af4f8c4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106898 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26616 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17013 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112107 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/57466 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/108902 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27284 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35118 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81179 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109867 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/21664 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96435 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61520 "Found 1 new API test failure: /WPE/TestSSL:/webkit/WebKitWebView/ephemeral-tls-errors (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21107 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14544 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56915 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91006 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14571 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115089 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34002 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25092 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90234 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34369 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92666 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89945 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22944 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34867 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12681 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/29708 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33926 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/39372 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33673 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35298 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
LGTM, please remember to export the WPT change :)

anttijk committed:
[popover] Integrate with anchor positioning
https://bugs.webkit.org/show_bug.cgi?id=284226
rdar://141094328

Reviewed by Alan Baradlay and Tim Nguyen.

See https://github.com/whatwg/html/pull/10728.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/popover-implicit-anchor-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/popover-implicit-anchor.html:

Test against the actual button position so button margins from appearance don't affect the results.

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::setInvokedPopover):

Invalidate the element style so the we update the isPopoverInvoker style bit.

* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::isPopoverInvoker const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setIsPopoverInvoker):

Add a bit. We use it to determine of a box is an implicit anchor.
Making this a style bit allows anchor registration and updates work correctly
without any new mechanisms.

* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::findImplicitAnchor):

The implicit anchor for a popover is its invoker.
Factor into a function.

(WebCore::Style::findLastAcceptableAnchorWithName):
(WebCore::Style::AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement):
(WebCore::Style::AnchorPositionEvaluator::isAnchor):
(WebCore::Style::AnchorPositionEvaluator::isImplicitAnchor):

A popover invoker is an implicit anchor.
Factor into a function.

* Source/WebCore/style/AnchorPositionEvaluator.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Set isPopoverInvoker bit for invokers.

Canonical link: https://commits.webkit.org/295995@main
